File: packages/go-sdk/http_client.go
package wusul

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"time"
)

// HTTPClient handles authenticated requests to the Wusul API
type HTTPClient struct {
	client       *http.Client
	accountID    string
	sharedSecret string
	baseURL      string
}

// NewHTTPClient creates a new HTTP client
func NewHTTPClient(accountID, sharedSecret, baseURL string, timeout time.Duration) *HTTPClient {
	return &HTTPClient{
		client: &http.Client{
			Timeout: timeout,
		},
		accountID:    accountID,
		sharedSecret: sharedSecret,
		baseURL:      baseURL,
	}
}

// Get makes a GET request
func (c *HTTPClient) Get(path string, sigPayload map[string]interface{}, result interface{}) error {
	headers, encodedPayload, err := createGetAuthHeaders(c.accountID, c.sharedSecret, sigPayload)
	if err != nil {
		return fmt.Errorf("failed to create auth headers: %w", err)
	}

	// Build URL with query parameter
	fullURL := c.baseURL + path
	if encodedPayload != "" {
		parsedURL, err := url.Parse(fullURL)
		if err != nil {
			return fmt.Errorf("failed to parse URL: %w", err)
		}
		q := parsedURL.Query()
		q.Set("sig_payload", encodedPayload)
		parsedURL.RawQuery = q.Encode()
		fullURL = parsedURL.String()
	}

	req, err := http.NewRequest("GET", fullURL, nil)
	if err != nil {
		return fmt.Errorf("failed to create request: %w", err)
	}

	for key, value := range headers {
		req.Header.Set(key, value)
	}

	return c.doRequest(req, result)
}

// Post makes a POST request
func (c *HTTPClient) Post(path string, data interface{}, result interface{}) error {
	headers, err := createAuthHeaders(c.accountID, c.sharedSecret, data)
	if err != nil {
		return fmt.Errorf("failed to create auth headers: %w", err)
	}

	var body []byte
	if data != nil {
		body, err = json.Marshal(data)
		if err != nil {
			return fmt.Errorf("failed to marshal request body: %w", err)
		}
	}

	req, err := http.NewRequest("POST", c.baseURL+path, bytes.NewBuffer(body))
	if err != nil {
		return fmt.Errorf("failed to create request: %w", err)
	}

	for key, value := range headers {
		req.Header.Set(key, value)
	}

	return c.doRequest(req, result)
}

// Patch makes a PATCH request
func (c *HTTPClient) Patch(path string, data interface{}, result interface{}) error {
	headers, err := createAuthHeaders(c.accountID, c.sharedSecret, data)
	if err != nil {
		return fmt.Errorf("failed to create auth headers: %w", err)
	}

	var body []byte
	if data != nil {
		body, err = json.Marshal(data)
		if err != nil {
			return fmt.Errorf("failed to marshal request body: %w", err)
		}
	}

	req, err := http.NewRequest("PATCH", c.baseURL+path, bytes.NewBuffer(body))
	if err != nil {
		return fmt.Errorf("failed to create request: %w", err)
	}

	for key, value := range headers {
		req.Header.Set(key, value)
	}

	return c.doRequest(req, result)
}

// Delete makes a DELETE request
func (c *HTTPClient) Delete(path string, result interface{}) error {
	headers, err := createAuthHeaders(c.accountID, c.sharedSecret, nil)
	if err != nil {
		return fmt.Errorf("failed to create auth headers: %w", err)
	}

	req, err := http.NewRequest("DELETE", c.baseURL+path, nil)
	if err != nil {
		return fmt.Errorf("failed to create request: %w", err)
	}

	for key, value := range headers {
		req.Header.Set(key, value)
	}

	return c.doRequest(req, result)
}

// doRequest executes the HTTP request and handles the response
func (c *HTTPClient) doRequest(req *http.Request, result interface{}) error {
	resp, err := c.client.Do(req)
	if err != nil {
		return fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return fmt.Errorf("failed to read response body: %w", err)
	}

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		var errorResp struct {
			Error   string `json:"error"`
			Message string `json:"message"`
		}
		if err := json.Unmarshal(body, &errorResp); err == nil {
			errorMsg := errorResp.Error
			if errorMsg == "" {
				errorMsg = errorResp.Message
			}
			if errorMsg != "" {
				return fmt.Errorf("Wusul API Error (%d): %s", resp.StatusCode, errorMsg)
			}
		}
		return fmt.Errorf("Wusul API Error (%d): %s", resp.StatusCode, string(body))
	}

	if result != nil {
		// Try to unmarshal as a response with data field
		var dataResp struct {
			Data json.RawMessage `json:"data"`
		}
		if err := json.Unmarshal(body, &dataResp); err == nil && len(dataResp.Data) > 0 {
			if err := json.Unmarshal(dataResp.Data, result); err != nil {
				return fmt.Errorf("failed to unmarshal data field: %w", err)
			}
		} else {
			// Otherwise unmarshal directly
			if err := json.Unmarshal(body, result); err != nil {
				return fmt.Errorf("failed to unmarshal response: %w", err)
			}
		}
	}

	return nil
}

File: packages/csharp-sdk/src/Wusul/Http/HttpClient.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Text;
using System.Threading.Tasks;
using Newtonsoft.Json;
using Wusul.Auth;
using Wusul.Models;

namespace Wusul.Http
{
    /// <summary>
    /// HTTP client for making authenticated requests to the Wusul API
    /// </summary>
    public class WusulHttpClient : IDisposable
    {
        private readonly System.Net.Http.HttpClient _httpClient;
        private readonly string _accountId;
        private readonly string _sharedSecret;
        private bool _disposed = false;

        /// <summary>
        /// Creates a new HTTP client instance
        /// </summary>
        /// <param name="accountId">Account ID</param>
        /// <param name="sharedSecret">Shared secret</param>
        /// <param name="baseUrl">Base URL for the API</param>
        /// <param name="timeout">Request timeout in milliseconds</param>
        public WusulHttpClient(string accountId, string sharedSecret, string baseUrl, int timeout)
        {
            _accountId = accountId;
            _sharedSecret = sharedSecret;

            _httpClient = new System.Net.Http.HttpClient
            {
                BaseAddress = new Uri(baseUrl),
                Timeout = TimeSpan.FromMilliseconds(timeout)
            };

            _httpClient.DefaultRequestHeaders.Add("Accept", "application/json");
        }

        /// <summary>
        /// Makes a GET request
        /// </summary>
        /// <typeparam name="T">Response type</typeparam>
        /// <param name="url">Request URL</param>
        /// <param name="sigPayload">Payload for signature (optional)</param>
        /// <returns>Response data</returns>
        public async Task<T> GetAsync<T>(string url, object? sigPayload = null)
        {
            var payload = sigPayload ?? new { id = "0" };
            var (headers, encodedPayload) = AuthenticationHelper.CreateAuthHeaders(_accountId, _sharedSecret, payload);

            var request = new HttpRequestMessage(HttpMethod.Get, url);
            foreach (var header in headers)
            {
                request.Headers.TryAddWithoutValidation(header.Key, header.Value);
            }

            // Add sig_payload as query parameter
            var separator = url.Contains("?") ? "&" : "?";
            var fullUrl = $"{url}{separator}sig_payload={System.Net.WebUtility.UrlEncode(encodedPayload)}";
            request.RequestUri = new Uri(_httpClient.BaseAddress, fullUrl);

            var response = await _httpClient.SendAsync(request);
            return await HandleResponseAsync<T>(response);
        }

        /// <summary>
        /// Makes a POST request
        /// </summary>
        /// <typeparam name="T">Response type</typeparam>
        /// <param name="url">Request URL</param>
        /// <param name="data">Request body data (optional)</param>
        /// <returns>Response data</returns>
        public async Task<T> PostAsync<T>(string url, object? data = null)
        {
            var payload = data ?? new { id = "0" };
            var (headers, _) = AuthenticationHelper.CreateAuthHeaders(_accountId, _sharedSecret, payload);

            var request = new HttpRequestMessage(HttpMethod.Post, url);
            foreach (var header in headers)
            {
                request.Headers.TryAddWithoutValidation(header.Key, header.Value);
            }

            if (data != null)
            {
                var jsonContent = JsonConvert.SerializeObject(data);
                request.Content = new StringContent(jsonContent, Encoding.UTF8, "application/json");
            }

            var response = await _httpClient.SendAsync(request);
            return await HandleResponseAsync<T>(response);
        }

        /// <summary>
        /// Makes a PATCH request
        /// </summary>
        /// <typeparam name="T">Response type</typeparam>
        /// <param name="url">Request URL</param>
        /// <param name="data">Request body data (optional)</param>
        /// <returns>Response data</returns>
        public async Task<T> PatchAsync<T>(string url, object? data = null)
        {
            var payload = data ?? new { id = "0" };
            var (headers, _) = AuthenticationHelper.CreateAuthHeaders(_accountId, _sharedSecret, payload);

            var request = new HttpRequestMessage(new HttpMethod("PATCH"), url);
            foreach (var header in headers)
            {
                request.Headers.TryAddWithoutValidation(header.Key, header.Value);
            }

            if (data != null)
            {
                var jsonContent = JsonConvert.SerializeObject(data);
                request.Content = new StringContent(jsonContent, Encoding.UTF8, "application/json");
            }

            var response = await _httpClient.SendAsync(request);
            return await HandleResponseAsync<T>(response);
        }

        /// <summary>
        /// Makes a DELETE request
        /// </summary>
        /// <typeparam name="T">Response type</typeparam>
        /// <param name="url">Request URL</param>
        /// <returns>Response data</returns>
        public async Task<T> DeleteAsync<T>(string url)
        {
            var (headers, _) = AuthenticationHelper.CreateAuthHeaders(_accountId, _sharedSecret);

            var request = new HttpRequestMessage(HttpMethod.Delete, url);
            foreach (var header in headers)
            {
                request.Headers.TryAddWithoutValidation(header.Key, header.Value);
            }

            var response = await _httpClient.SendAsync(request);
            return await HandleResponseAsync<T>(response);
        }

        /// <summary>
        /// Handles the HTTP response and deserializes the result
        /// </summary>
        private async Task<T> HandleResponseAsync<T>(HttpResponseMessage response)
        {
            var content = await response.Content.ReadAsStringAsync();

            if (!response.IsSuccessStatusCode)
            {
                var errorMessage = content;
                try
                {
                    var errorResponse = JsonConvert.DeserializeObject<WusulResponse<object>>(content);
                    errorMessage = errorResponse?.Error ?? errorResponse?.Message ?? content;
                }
                catch
                {
                    // If we can't parse the error, use the raw content
                }

                throw new WusulException($"Wusul API Error ({(int)response.StatusCode}): {errorMessage}");
            }

            try
            {
                var wusulResponse = JsonConvert.DeserializeObject<WusulResponse<T>>(content);
                if (wusulResponse != null && !EqualityComparer<T>.Default.Equals(wusulResponse.Data, default(T)))
                {
                    return wusulResponse.Data;
                }

                // If there's no data wrapper, try to deserialize directly
                var directResponse = JsonConvert.DeserializeObject<T>(content);
                if (directResponse != null && !EqualityComparer<T>.Default.Equals(directResponse, default(T)))
                {
                    return directResponse;
                }

                throw new WusulException("Response data is null");
            }
            catch (JsonException ex)
            {
                throw new WusulException($"Failed to deserialize response: {ex.Message}");
            }
        }

        /// <summary>
        /// Disposes the HTTP client
        /// </summary>
        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }

        /// <summary>
        /// Disposes the HTTP client
        /// </summary>
        protected virtual void Dispose(bool disposing)
        {
            if (!_disposed)
            {
                if (disposing)
                {
                    _httpClient?.Dispose();
                }
                _disposed = true;
            }
        }
    }

    /// <summary>
    /// Exception thrown when a Wusul API error occurs
    /// </summary>
    public class WusulException : Exception
    {
        public WusulException(string message) : base(message)
        {
        }

        public WusulException(string message, Exception innerException) : base(message, innerException)
        {
        }
    }
}
